Flaky TestRuler_Rules

As seen at https://github.com/grafana/mimir/runs/7956562047
Re-run at https://github.com/grafana/mimir/runs/7956932204 passed.

```
--- FAIL: TestRuler_Rules (0.06s)
    --- FAIL: TestRuler_Rules/rules_-_user1 (0.02s)
        ruler_test.go:283: 
            	Error Trace:	/__w/mimir/mimir/pkg/ruler/ruler_test.go:283
            	Error:      	"[]" should have 1 item(s), but has 0
            	Test:       	TestRuler_Rules/rules_-_user1
```

Just seen again [here](https://github.com/grafana/mimir/actions/runs/3105675113/jobs/5031576928):

```
--- FAIL: TestRuler_Rules (0.02s)
    --- FAIL: TestRuler_Rules/federated_rule_group (0.00s)
        ruler_test.go:283: 
            	Error Trace:	/__w/mimir/mimir/pkg/ruler/ruler_test.go:283
            	Error:      	"[]" should have 1 item(s), but has 0
            	Test:       	TestRuler_Rules/federated_rule_group
```

--- a/pkg/ruler/api_test.go
+++ b/pkg/ruler/api_test.go
@@ -14,10 +14,12 @@ import (
 	"net/http/httptest"
 	"strings"
 	"testing"
+	"time"
 
 	"github.com/go-kit/log"
 	"github.com/gorilla/mux"
 	"github.com/grafana/dskit/services"
+	"github.com/grafana/dskit/test"
 	"github.com/stretchr/testify/require"
 	"github.com/weaveworks/common/user"
 
@@ -26,9 +28,8 @@ import (
 
 func TestRuler(t *testing.T) {
 	testCases := map[string]struct {
-		mockRules map[string]rulespb.RuleGroupList
-		userID    string
-
+		mockRules        map[string]rulespb.RuleGroupList
+		userID           string
 		expectedResponse response
 	}{
 		"rules": {
@@ -159,13 +160,20 @@ func TestRuler(t *testing.T) {
 
 			r := buildRuler(t, cfg, newMockRuleStore(tc.mockRules), rulerAddrMap)
 			require.NoError(t, services.StartAndAwaitRunning(context.Background(), r))
-			defer services.StopAndAwaitTerminated(context.Background(), r) //nolint:errcheck
+			t.Cleanup(func() {
+				require.NoError(t, services.StopAndAwaitTerminated(context.Background(), r))
+			})
 
 			// Make sure mock grpc client can find this instance, based on instance address registered in the ring.
 			rulerAddrMap[r.lifecycler.GetInstanceAddr()] = r
 
-			// Ensure all rules are loaded before usage
-			r.syncRules(context.Background(), rulerSyncReasonInitial)
+			// Rules will be synchronized asynchronously, so we wait until the expected number of rule groups
+			// has been synched.
+			test.Poll(t, 5*time.Second, len(tc.mockRules[tc.userID]), func() interface{} {
+				ctx := user.InjectOrgID(context.Background(), tc.userID)
+				rls, _ := r.Rules(ctx, &RulesRequest{})
+				return len(rls.Groups)
+			})
 
 			a := NewAPI(r, r.store, log.NewNopLogger())
 
@@ -199,13 +207,20 @@ func TestRuler_alerts(t *testing.T) {
 
 	r := buildRuler(t, cfg, newMockRuleStore(mockRules), rulerAddrMap)
 	require.NoError(t, services.StartAndAwaitRunning(context.Background(), r))
-	defer services.StopAndAwaitTerminated(context.Background(), r) //nolint:errcheck
+	t.Cleanup(func() {
+		require.NoError(t, services.StopAndAwaitTerminated(context.Background(), r))
+	})
 
 	// Make sure mock grpc client can find this instance, based on instance address registered in the ring.
 	rulerAddrMap[r.lifecycler.GetInstanceAddr()] = r
 
-	// Ensure all rules are loaded before usage
-	r.syncRules(context.Background(), rulerSyncReasonInitial)
+	// Rules will be synchronized asynchronously, so we wait until the expected number of rule groups
+	// has been synched.
+	test.Poll(t, 5*time.Second, len(mockRules["user1"]), func() interface{} {
+		ctx := user.InjectOrgID(context.Background(), "user1")
+		rls, _ := r.Rules(ctx, &RulesRequest{})
+		return len(rls.Groups)
+	})
 
 	a := NewAPI(r, r.store, log.NewNopLogger())
 
@@ -238,9 +253,7 @@ func TestRuler_alerts(t *testing.T) {
 func TestRuler_Create(t *testing.T) {
 	cfg := defaultRulerConfig(t)
 
-	r := newTestRuler(t, cfg, newMockRuleStore(make(map[string]rulespb.RuleGroupList)))
-	defer services.StopAndAwaitTerminated(context.Background(), r) //nolint:errcheck
-
+	r := buildAndStartRuler(t, cfg, newMockRuleStore(make(map[string]rulespb.RuleGroupList)))
 	a := NewAPI(r, r.store, log.NewNopLogger())
 
 	tc := []struct {
@@ -366,9 +379,7 @@ func TestRuler_DeleteNamespace(t *testing.T) {
 		},
 	}
 
-	r := newTestRuler(t, cfg, newMockRuleStore(mockRulesNamespaces))
-	defer services.StopAndAwaitTerminated(context.Background(), r) //nolint:errcheck
-
+	r := buildAndStartRuler(t, cfg, newMockRuleStore(mockRulesNamespaces))
 	a := NewAPI(r, r.store, log.NewNopLogger())
 
 	router := mux.NewRouter()
@@ -403,9 +414,7 @@ func TestRuler_DeleteNamespace(t *testing.T) {
 func TestRuler_LimitsPerGroup(t *testing.T) {
 	cfg := defaultRulerConfig(t)
 
-	r := newTestRuler(t, cfg, newMockRuleStore(make(map[string]rulespb.RuleGroupList)))
-	defer services.StopAndAwaitTerminated(context.Background(), r) //nolint:errcheck
-
+	r := buildAndStartRuler(t, cfg, newMockRuleStore(make(map[string]rulespb.RuleGroupList)))
 	r.limits = &ruleLimits{maxRuleGroups: 1, maxRulesPerRuleGroup: 1}
 
 	a := NewAPI(r, r.store, log.NewNopLogger())
@@ -456,9 +465,7 @@ rules:
 func TestRuler_RulerGroupLimits(t *testing.T) {
 	cfg := defaultRulerConfig(t)
 
-	r := newTestRuler(t, cfg, newMockRuleStore(make(map[string]rulespb.RuleGroupList)))
-	defer services.StopAndAwaitTerminated(context.Background(), r) //nolint:errcheck
-
+	r := buildAndStartRuler(t, cfg, newMockRuleStore(make(map[string]rulespb.RuleGroupList)))
 	r.limits = &ruleLimits{maxRuleGroups: 1, maxRulesPerRuleGroup: 1}
 
 	a := NewAPI(r, r.store, log.NewNopLogger())

--- a/pkg/ruler/manager.go
+++ b/pkg/ruler/manager.go
@@ -98,6 +98,8 @@ func NewDefaultMultiTenantManager(cfg Config, managerFactory ManagerFactory, reg
 	}, nil
 }
 
+// SyncRuleGroups sync the input rulesGroups.
+// It's not safe to call this function concurrently.
 func (r *DefaultMultiTenantManager) SyncRuleGroups(ctx context.Context, ruleGroups map[string]rulespb.RuleGroupList) {
 	if !r.cfg.TenantFederation.Enabled {
 		RemoveFederatedRuleGroups(ruleGroups)

--- a/pkg/ruler/ruler.go
+++ b/pkg/ruler/ruler.go
@@ -459,6 +459,7 @@ func (r *Ruler) run(ctx context.Context) error {
 }
 
 // It's not safe to call this function concurrently.
+// We expect this function is only called from Ruler.run().
 func (r *Ruler) syncRules(ctx context.Context, reason string) {
 	level.Debug(r.logger).Log("msg", "syncing rules", "reason", reason)
 	r.metrics.rulerSync.WithLabelValues(reason).Inc()

--- a/pkg/ruler/ruler.go
+++ b/pkg/ruler/ruler.go
@@ -458,6 +458,7 @@ func (r *Ruler) run(ctx context.Context) error {
 	}
 }
 
+// It's not safe to call this function concurrently.
 func (r *Ruler) syncRules(ctx context.Context, reason string) {
 	level.Debug(r.logger).Log("msg", "syncing rules", "reason", reason)
 	r.metrics.rulerSync.WithLabelValues(reason).Inc()

--- a/pkg/ruler/ruler_test.go
+++ b/pkg/ruler/ruler_test.go
@@ -27,6 +27,7 @@ import (
 	"github.com/grafana/dskit/kv/consul"
 	"github.com/grafana/dskit/ring"
 	"github.com/grafana/dskit/services"
+	"github.com/grafana/dskit/test"
 	"github.com/prometheus/client_golang/prometheus"
 	prom_testutil "github.com/prometheus/client_golang/prometheus/testutil"
 	"github.com/prometheus/prometheus/model/labels"
@@ -177,12 +178,14 @@ func buildRuler(t *testing.T, cfg Config, storage rulestore.RuleStore, rulerAddr
 	return ruler
 }
 
-func newTestRuler(t *testing.T, cfg Config, storage rulestore.RuleStore) *Ruler {
+func buildAndStartRuler(t *testing.T, cfg Config, storage rulestore.RuleStore) *Ruler {
 	ruler := buildRuler(t, cfg, storage, nil)
 	require.NoError(t, services.StartAndAwaitRunning(context.Background(), ruler))
 
-	// Ensure all rules are loaded before usage
-	ruler.syncRules(context.Background(), rulerSyncReasonInitial)
+	// Ensure the service is stopped at the end of the test.
+	t.Cleanup(func() {
+		require.NoError(t, services.StopAndAwaitTerminated(context.Background(), ruler))
+	})
 
 	return ruler
 }
@@ -274,10 +277,16 @@ func TestRuler_Rules(t *testing.T) {
 			cfg := defaultRulerConfig(t)
 			cfg.TenantFederation.Enabled = true
 
-			r := newTestRuler(t, cfg, newMockRuleStore(tc.mockRules))
-			defer services.StopAndAwaitTerminated(context.Background(), r) //nolint:errcheck
+			r := buildAndStartRuler(t, cfg, newMockRuleStore(tc.mockRules))
 
+			// Rules will be synchronized asynchronously, so we wait until the expected number of rule groups
+			// has been synched.
 			ctx := user.InjectOrgID(context.Background(), tc.userID)
+			test.Poll(t, 5*time.Second, len(mockRules[tc.userID]), func() interface{} {
+				rls, _ := r.Rules(ctx, &RulesRequest{})
+				return len(rls.Groups)
+			})
+
 			rls, err := r.Rules(ctx, &RulesRequest{})
 			require.NoError(t, err)
 			require.Len(t, rls.Groups, len(mockRules[tc.userID]))
@@ -1003,8 +1012,7 @@ type ruleGroupKey struct {
 func TestRuler_ListAllRules(t *testing.T) {
 	cfg := defaultRulerConfig(t)
 
-	r := newTestRuler(t, cfg, newMockRuleStore(mockRules))
-	defer services.StopAndAwaitTerminated(context.Background(), r) //nolint:errcheck
+	r := buildAndStartRuler(t, cfg, newMockRuleStore(mockRules))
 
 	router := mux.NewRouter()
 	router.Path("/ruler/rule_groups").Methods(http.MethodGet).HandlerFunc(r.ListAllRules)

